pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java

```diff
@@ -0,0 +1,195 @@
+      _valueBuffers[valueBufferIndex] = valueBuffer;
+      _totalBufferSize += VALUE_BUFFER_SIZE;
+      _valueStartOffset = valueBufferIndex << VALUE_BUFFER_SIFT_OFFSET;
+    } else {
+      valueBuffer = _valueBuffers[valueBufferIndex];
+    }
+
+    int valueEndOffset = _valueStartOffset + valueLength;
+    offsetBuffer.putInt((offsetIndex + 1) << 2, valueEndOffset);
+    valueBuffer.readFrom(_valueStartOffset & VALUE_BUFFER_MASK, value);
+    _valueStartOffset = valueEndOffset;
+
+    return _numValues++;
+  }
+
+  public byte[] get(int index) {
+    int offsetBufferIndex = index >>> OFFSET_BUFFER_SHIFT_OFFSET;
+    int offsetIndex = index & OFFSET_BUFFER_MASK;
+    int previousValueEndOffset = _offsetBuffers[offsetBufferIndex].getInt(offsetIndex << 2);
+    int valueEndOffset = _offsetBuffers[offsetBufferIndex].getInt((offsetIndex + 1) << 2);
+
+    if (previousValueEndOffset == valueEndOffset) {
+      return EMPTY_BYTES;
+    }
+
+    int valueBufferIndex = (valueEndOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
+    int startOffsetInValueBuffer;
+    int valueLength;
+    if ((previousValueEndOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET != valueBufferIndex) {
+      // The first value in the value buffer
+      startOffsetInValueBuffer = 0;
+      valueLength = valueEndOffset & VALUE_BUFFER_MASK;
+    } else {
+      // Not the first value in the value buffer
+      startOffsetInValueBuffer = previousValueEndOffset & VALUE_BUFFER_MASK;
+      valueLength = valueEndOffset - previousValueEndOffset;
+    }
+
+    byte[] value = new byte[valueLength];
+    _valueBuffers[valueBufferIndex].copyTo(startOffsetInValueBuffer, value);
+    return value;
+  }
+
+  public boolean equals(int index, byte[] value) {
+    int valueLength = value.length;
+
+    int offsetBufferIndex = index >>> OFFSET_BUFFER_SHIFT_OFFSET;
+    int offsetIndex = index & OFFSET_BUFFER_MASK;
+    int previousValueEndOffset = _offsetBuffers[offsetBufferIndex].getInt(offsetIndex << 2);
+    int valueEndOffset = _offsetBuffers[offsetBufferIndex].getInt((offsetIndex + 1) << 2);
+
+    if (valueLength == 0) {
+      return previousValueEndOffset == valueEndOffset;
+    }
+
+    int valueBufferIndex = (valueEndOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
+    int startOffsetInValueBuffer;
+    if ((previousValueEndOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET != valueBufferIndex) {
+      // The first value in the value buffer
+      if ((valueEndOffset & VALUE_BUFFER_MASK) != valueLength) {
+        return false;
+      }
+      startOffsetInValueBuffer = 0;
+    } else {
+      // Not the first value in the value buffer
+      if (valueEndOffset - previousValueEndOffset != valueLength) {
+        return false;
+      }
+      startOffsetInValueBuffer = previousValueEndOffset & VALUE_BUFFER_MASK;
+    }
+
+    byte[] valueAtIndex = new byte[valueLength];
+    _valueBuffers[valueBufferIndex].copyTo(startOffsetInValueBuffer, valueAtIndex);
```

Comment:
Wouldnt it be faster if you just compared by fetching the bytes from the valueBuffer?

Reply:
You mean byte-by-byte comparison? If we reach here, that means there is a pretty high chance that the two values are the same (we call this method when hash code matches, and we can only reach here if the length also matches), thus doing batch read should be preferable. I added one extra check to optimize the case of small reads to compare byte-by-byte.

Reply:
But then there is extra garbage generated in the read.

Reply:
True, but that is short lived garbage, and the benchmark in #4362 already proved that batch read can give big performance gain for big values.